File: tests/language-support/haskell/src/Storage.hs
-- | Storage operations for managing books on shelves.
module Storage
    ( addBook
    , removeBook
    , findByAuthor
    , findByYearRange
    , sortByTitle
    , isFull
    , generateReport
    ) where

import Data.Char (toLower)
import Data.List (sortBy, isInfixOf, intercalate, nub)
import Models

-- | Add a book to the shelf. Returns Left on error.
addBook :: Shelf -> Book -> Either String Shelf
addBook shelf book
    | isFull shelf = Left $ "Shelf '" ++ shelfName shelf ++ "' is at capacity (" ++ show (shelfCapacity shelf) ++ ")"
    | otherwise    = Right shelf { shelfBooks = shelfBooks shelf ++ [book] }

-- | Remove a book by ISBN. Returns Left on error.
removeBook :: Shelf -> String -> Either String (Shelf, Book)
removeBook shelf isbn =
    case break (\b -> bookIsbn b == isbn) (shelfBooks shelf) of
        (_, [])    -> Left $ "No book with ISBN '" ++ isbn ++ "' found on shelf"
        (before, b:after) -> Right (shelf { shelfBooks = before ++ after }, b)

-- | Find books by author (case-insensitive substring match).
findByAuthor :: Shelf -> String -> [Book]
findByAuthor shelf query =
    filter (\b -> map toLower query `isInfixOf` map toLower (bookAuthor b)) (shelfBooks shelf)

-- | Find books within an inclusive year range.
findByYearRange :: Shelf -> Int -> Int -> [Book]
findByYearRange shelf startYear endYear =
    filter (\b -> bookYear b >= startYear && bookYear b <= endYear) (shelfBooks shelf)

-- | Return books sorted by title without mutating.
sortByTitle :: Shelf -> [Book]
sortByTitle shelf = sortBy (\a b -> compare (bookTitle a) (bookTitle b)) (shelfBooks shelf)

-- | Check whether the shelf is at capacity.
isFull :: Shelf -> Bool
isFull shelf = length (shelfBooks shelf) >= shelfCapacity shelf

-- | Generate a formatted report for the shelf.
generateReport :: Shelf -> String
generateReport shelf =
    let books = shelfBooks shelf
        total = length books
        availCount = length (filter bookAvailable books)
        availPct = if total > 0 then (availCount * 100) `div` total else 0
        capPct = if shelfCapacity shelf > 0 then (total * 100) `div` shelfCapacity shelf else 0
        authors = map bookAuthor books
        uniqueAuthors = nub authors
        uniqueCount = length uniqueAuthors
        authorCounts = map (\a -> (a, length (filter (== a) authors))) uniqueAuthors
        years = map bookYear books
        minYear = minimum years
        maxYear = maximum years
        header = "=== Library Report: " ++ shelfName shelf ++ " ==="
        bookLines = map formatAvailLine books
        authorLines = map (\(a, c) -> "  - " ++ a ++ " (" ++ show c ++ " books)") authorCounts
    in intercalate "\n" $
        [ header
        , "Total books: " ++ show total
        , "Available: " ++ show availCount ++ " / " ++ show total ++ " (" ++ show availPct ++ "%)"
        , "Capacity: " ++ show total ++ " / " ++ show (shelfCapacity shelf) ++ " (" ++ show capPct ++ "% full)"
        , ""
        , "Authors (" ++ show uniqueCount ++ " unique):"
        ] ++ authorLines ++
        [ ""
        , "Year range: " ++ show minYear ++ " - " ++ show maxYear
        , ""
        , "Books by availability:"
        ] ++ bookLines

formatAvailLine :: Book -> String
formatAvailLine book =
    let marker = if bookAvailable book then "+" else "-"
    in "  [" ++ marker ++ "] " ++ bookTitle book ++ " by " ++ bookAuthor book
       ++ " (" ++ show (bookYear book) ++ ") - " ++ bookIsbn book

File: tests/language-support/ocaml/lib/storage.ml
(** Storage operations for managing books on shelves. *)

open Models

(** Check whether the shelf is at capacity. *)
let is_full shelf = List.length shelf.books >= shelf.capacity

(** Add a book to the shelf. *)
let add_book shelf book =
  if is_full shelf then
    Error (Printf.sprintf "Shelf '%s' is at capacity (%d)" shelf.name shelf.capacity)
  else begin
    shelf.books <- shelf.books @ [book];
    Ok ()
  end

(** Remove a book by ISBN. *)
let remove_book shelf isbn =
  let rec aux acc = function
    | [] -> Error (Printf.sprintf "No book with ISBN '%s' found on shelf" isbn)
    | b :: rest when b.isbn = isbn ->
      shelf.books <- List.rev acc @ rest;
      Ok b
    | b :: rest -> aux (b :: acc) rest
  in
  aux [] shelf.books

(** Find books by author (case-insensitive substring match). *)
let find_by_author shelf query =
  let lq = String.lowercase_ascii query in
  List.filter (fun b ->
    let la = String.lowercase_ascii b.author in
    let qlen = String.length lq in
    let alen = String.length la in
    if qlen > alen then false
    else
      let found = ref false in
      for i = 0 to alen - qlen do
        if String.sub la i qlen = lq then found := true
      done;
      !found
  ) shelf.books

(** Find books within an inclusive year range. *)
let find_by_year_range shelf start_year end_year =
  List.filter (fun b -> b.year >= start_year && b.year <= end_year) shelf.books

(** Return books sorted by title without mutating. *)
let sort_by_title shelf =
  List.sort (fun a b -> String.compare a.title b.title) shelf.books

(** Count occurrences preserving insertion order of unique elements. *)
let count_authors books =
  let seen = ref [] in
  List.iter (fun b ->
    if not (List.mem b.author (List.map fst !seen)) then
      seen := !seen @ [(b.author, List.length (List.filter (fun x -> x.author = b.author) books))]
  ) books;
  !seen

(** Generate a formatted report for the shelf. *)
let generate_report shelf =
  let books = shelf.books in
  let total = List.length books in
  let avail_count = List.length (List.filter (fun b -> b.available) books) in
  let avail_pct = if total > 0 then (avail_count * 100) / total else 0 in
  let cap_pct = if shelf.capacity > 0 then (total * 100) / shelf.capacity else 0 in
  let author_counts = count_authors books in
  let unique_count = List.length author_counts in
  let years = List.map (fun b -> b.year) books in
  let min_year = List.fold_left min max_int years in
  let max_year = List.fold_left max min_int years in
  let buf = Buffer.create 512 in
  let add s = Buffer.add_string buf s in
  let addln s = add s; add "\n" in
  addln (Printf.sprintf "=== Library Report: %s ===" shelf.name);
  addln (Printf.sprintf "Total books: %d" total);
  addln (Printf.sprintf "Available: %d / %d (%d%%)" avail_count total avail_pct);
  addln (Printf.sprintf "Capacity: %d / %d (%d%% full)" total shelf.capacity cap_pct);
  addln "";
  addln (Printf.sprintf "Authors (%d unique):" unique_count);
  List.iter (fun (a, c) ->
    addln (Printf.sprintf "  - %s (%d books)" a c)
  ) author_counts;
  addln "";
  addln (Printf.sprintf "Year range: %d - %d" min_year max_year);
  addln "";
  addln "Books by availability:";
  List.iter (fun b ->
    let marker = if b.available then "+" else "-" in
    addln (Printf.sprintf "  [%s] %s by %s (%d) - %s" marker b.title b.author b.year b.isbn)
  ) books;
  (* Remove trailing newline *)
  let s = Buffer.contents buf in
  if String.length s > 0 && s.[String.length s - 1] = '\n' then
    String.sub s 0 (String.length s - 1)
  else
    s
